// @1inch/solidity-utils/dist/typechain-types/factories/@0xsequence/create3/contracts/Create3__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Create3__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "ErrorCreatingContract",
        type: "error",
    },
    {
        inputs: [],
        name: "ErrorCreatingProxy",
        type: "error",
    },
    {
        inputs: [],
        name: "TargetAlreadyExists",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea264697066735822122099deb8fbccc0ed640c2ccaa57f93cfb12f791290c0a8032ea47dfc8b4ece489664736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class Create3__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.Create3__factory = Create3__factory;
Create3__factory.bytecode = _bytecode;
Create3__factory.abi = _abi;
