// @1inch/solidity-utils/dist/typechain-types/factories/@openzeppelin/contracts/utils/Address__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Address__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "address",
                name: "target",
                type: "address",
            },
        ],
        name: "AddressEmptyCode",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea26469706673582212200de9151a7ca92dc0c01f7742f7fd05d1d9dfb59fa6f40374841ccade5fe3760d64736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class Address__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.Address__factory = Address__factory;
Address__factory.bytecode = _bytecode;
Address__factory.abi = _abi;
