// @1inch/solidity-utils/dist/typechain-types/factories/@openzeppelin/contracts/utils/ShortStrings__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShortStrings__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "InvalidShortString",
        type: "error",
    },
    {
        inputs: [
            {
                internalType: "string",
                name: "str",
                type: "string",
            },
        ],
        name: "StringTooLong",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea2646970667358221220a56593ec449aed6bdc8c46593219041952db0f3fcd29f2f92c4e774ffa29e47864736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class ShortStrings__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.ShortStrings__factory = ShortStrings__factory;
ShortStrings__factory.bytecode = _bytecode;
ShortStrings__factory.abi = _abi;
