// @1inch/solidity-utils/dist/typechain-types/factories/contracts/libraries/BySigTraits__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BySigTraits__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "WrongNonceType",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea26469706673582212207a1e7584828109d42af34284144b3da003bc43413a63aa9540b923578e9c518b64736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class BySigTraits__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.BySigTraits__factory = BySigTraits__factory;
BySigTraits__factory.bytecode = _bytecode;
BySigTraits__factory.abi = _abi;
