// @1inch/solidity-utils/dist/typechain-types/factories/contracts/libraries/BytesStorage__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BytesStorage__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "OutOfBounds",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea2646970667358221220a5c60007d67618c0eb80c69c6060e3480789f6dfac95ae54c7cdab8fc5cf86d864736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class BytesStorage__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.BytesStorage__factory = BytesStorage__factory;
BytesStorage__factory.bytecode = _bytecode;
BytesStorage__factory.abi = _abi;
