// @1inch/solidity-utils/dist/typechain-types/factories/contracts/libraries/BytesMemory__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BytesMemory__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "OutOfBounds",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea26469706673582212208d9aa042252606f6917810907fe1ffed4a04578c7ead60f4408b176ce80dd1a664736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class BytesMemory__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.BytesMemory__factory = BytesMemory__factory;
BytesMemory__factory.bytecode = _bytecode;
BytesMemory__factory.abi = _abi;
