// @1inch/solidity-utils/dist/typechain-types/factories/contracts/libraries/UniERC20__factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UniERC20__factory = void 0;
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [],
        name: "ApproveCalledOnETH",
        type: "error",
    },
    {
        inputs: [],
        name: "ETHTransferFailed",
        type: "error",
    },
    {
        inputs: [],
        name: "FromIsNotSender",
        type: "error",
    },
    {
        inputs: [],
        name: "InsufficientBalance",
        type: "error",
    },
    {
        inputs: [],
        name: "NotEnoughValue",
        type: "error",
    },
    {
        inputs: [],
        name: "ToIsNotThis",
        type: "error",
    },
];
const _bytecode = "0x6080806040523460175760399081601c823930815050f35b5f80fdfe5f80fdfea2646970667358221220f0b31e4dcb53ac4ca20ccf87fa32089f049ea88f05641c4c9a0719a80361664e64736f6c63430008190033";
const isSuperArgs = (xs) => xs.length > 1;
class UniERC20__factory extends ethers_1.ContractFactory {
    constructor(...args) {
        if (isSuperArgs(args)) {
            super(...args);
        }
        else {
            super(_abi, _bytecode, args[0]);
        }
    }
    getDeployTransaction(overrides) {
        return super.getDeployTransaction(overrides || {});
    }
    deploy(overrides) {
        return super.deploy(overrides || {});
    }
    connect(runner) {
        return super.connect(runner);
    }
    static createInterface() {
        return new ethers_1.Interface(_abi);
    }
    static connect(address, runner) {
        return new ethers_1.Contract(address, _abi, runner);
    }
}
exports.UniERC20__factory = UniERC20__factory;
UniERC20__factory.bytecode = _bytecode;
UniERC20__factory.abi = _abi;
